### PR TITLE
Fix container hostPid settings

### LINKED ISSUE #44041: hostPID doesn't seem to work on Kubernetes 1.6
*Kubernetes version** (use `kubectl version`):
1.6.1, 1.6.0 (both run into this issue)

**Environment**:
- **Cloud provider or hardware configuration**: VirtualBox 5.1.18 VM
- **OS**: 16.04.2 LTS (Xenial Xerus)
- **Kernel**: `4.4.0-71-generic #92-Ubuntu SMP Fri Mar 24 12:59:01 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux` 
- **Install tools**: kubeadm

**What happened**:
A pod scheduled using hostPID wasn't using the host's PID namespace.

I've investigated this issue together with @neelimamukiri.

**What you expected to happen**:
I expected `"PidMode": "host",` in the Docker container's config.
Instead, the scheduled pod's container had `"PidMode": "",`

**How to reproduce it**:
Use this config to get a pod running a Docker container:
```
---

apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: busybox
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: busybox
subjects:
- kind: ServiceAccount
  name: busybox
  namespace: kube-system
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: busybox
  namespace: kube-system
rules:
  - apiGroups:
      - '*'
    resources:
      - '*'
    verbs:
      - '*'
  - apiGroups: [""]
    resources:
      - '*'
    verbs:
      - '*'
  - apiGroups: ["extensions"]
    resources:
      - '*'
    verbs:
      - '*'
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: busybox
  namespace: kube-system

---


kind: DaemonSet
apiVersion: extensions/v1beta1
metadata:
  name: busybox
  namespace: kube-system
  labels:
    k8s-app: busybox
spec:
  selector:
    matchLabels:
      k8s-app: busybox
  template:
    metadata:
      labels:
        k8s-app: busybox
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
    spec:
      hostNetwork: true
      hostPID: true
      tolerations:
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
      serviceAccountName: busybox
      containers:
        - name: busybox
          image: busybox
          command:
            - sleep
            - "3600"
          securityContext:
            privileged: true
--- 
```
Use `docker inspect` on the running container and check whether PidMode is set to host or not.

**Anything else we need to know**:
This was working properly on Kubernetes 1.5.3.

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

HostPid is not set correctly for containers.

**Which issue this PR fixes**

Fixes #44041.

**Special notes for your reviewer**:

Should be cherry-picked into v1.6 branch.

**Release note**:

```release-note
Fix container hostPid settings.
```

cc @yujuhong @derekwaynecarr @unclejack @kubernetes/sig-node-pr-reviews

### CODE REVIEW

**Comment by bboreham** [comment]
Do tests need improving to catch this?

**Comment by feiskyer** [comment]
> Do tests need improving to catch this?

I think so. We should add hostnetwork/hostpid testings to node e2e tests. I will add them in a separate PR.

**Comment by derekwaynecarr** [comment]
would like a node e2e test before merge.

File: pkg/kubelet/dockershim/security_context_test.go [hunk 1]
```diff
@@ -306,6 +306,7 @@ func TestModifyContainerNamespaceOptions(t *testing.T) {
 			expected: &dockercontainer.HostConfig{
 				NetworkMode: dockercontainer.NetworkMode(sandboxNSMode),
 				IpcMode:     dockercontainer.IpcMode(sandboxNSMode),
+				PidMode:     namespaceModeHost,
```
Comment (yujuhong) [inline pkg/kubelet/dockershim/security_context_test.go]: Does the test fail without the fix?

Comment (feiskyer) [inline pkg/kubelet/dockershim/security_context_test.go]: Yes

**Comment by yujuhong** [comment]
> would like a node e2e test before merge.

A node e2e test would be good.

On the other hand, we do have security context tests, but they are not running anywhere in our builds: https://github.com/kubernetes/kubernetes/issues/16456#issuecomment-283516201
@derekwaynecarr @pmorie, I thought OpenShift ran these tests. Did it not catch this bug?

**Comment by derekwaynecarr** [comment]
@yujuhong we are still rebasing on top Kubernetes 1.6, it's possible that it was not caught yet.

**Comment by yujuhong** [comment]
> @yujuhong we are still rebasing on top Kubernetes 1.6, it's possible that it was not caught yet.

:-(

**Comment by ncdc** [comment]
@sttts we'll need to pick this

**Comment by derekwaynecarr** [comment]
I manually verified the fix as well and things worked well.

**Pull request merged**